arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp

```diff
@@ -156,76 +177,125 @@ auto DocumentLeaderState::recoverEntries(std::unique_ptr<EntryIterator> ptr)
                                                   self->gid.database);
       } catch (...) {
         LOG_CTX("894f1", WARN, self->loggerContext)
-            << "failed to abort active transaction " << self->gid
+            << "failed to abort active transaction " << tid
             << " during recovery";
       }
     }
 
+    auto abortAllResult = transactionHandler->applyEntry(abortAll);
+    TRI_ASSERT(abortAllResult.ok());
+
+    self->release(abortAllRes.get());
+
     return {TRI_ERROR_NO_ERROR};
   });
 }
 
-auto DocumentLeaderState::replicateOperation(velocypack::SharedSlice payload,
-                                             OperationType operation,
-                                             TransactionId transactionId,
-                                             ShardID shard,
-                                             ReplicationOptions opts)
-    -> futures::Future<LogIndex> {
-  if (_isResigning.load()) {
-    LOG_CTX("ffe2f", INFO, loggerContext)
-        << "replicateOperation called on a resigning leader, will not "
-           "replicate";
-    return LogIndex{};  // TODO - can we do this differently instead of
-                        // returning a dummy index
-  }
-
-  TRI_ASSERT(operation != OperationType::kAbortAllOngoingTrx);
-  if ((operation == OperationType::kCommit ||
-       operation == OperationType::kAbort) &&
-      !_activeTransactions.getLockedGuard()->erase(transactionId)) {
-    // we have not replicated anything for a transaction with this id, so
-    // there is no need to replicate the abort/commit operation
-    return LogIndex{};  // TODO - can we do this differently instead of
-                        // returning a dummy index
-  }
+auto DocumentLeaderState::needsReplication(ReplicatedOperation const& op)
+    -> bool {
+  return std::visit(
+      overload{
+          [&](FinishesUserTransactionOrIntermediate auto const& op) -> bool {
+            // We don't replicate single abort/commit operations
+            // in case we have not replicated anything else for
+            // a transaction.
```

Comment:
This comment was confusing. Is it about not replicating the commit or abort for an empty transaction?

Reply:
Updated the comment